bot/utils/decorators.py

```diff
@@ -0,0 +1,321 @@
+
+                if current_month in allowed_months:
+                    await task_body(*args, **kwargs)
+                else:
+                    log.debug(f"Seasonal task {task_body.__qualname__} sleeps in {current_month.name}")
+
+                await asyncio.sleep(sleep_time)
+        return decorated_task
+    return decorator
+
+
+def in_month_listener(*allowed_months: Month) -> t.Callable:
+    """
+    Shield a listener from being invoked outside of `allowed_months`.
+
+    The check is performed against current UTC month.
+    """
+    def decorator(listener: t.Callable) -> t.Callable:
+        @functools.wraps(listener)
+        async def guarded_listener(*args, **kwargs) -> None:
+            """Wrapped listener will abort if not in allowed month."""
+            current_month = resolve_current_month()
+
+            if current_month in allowed_months:
+                # Propagate return value although it should always be None
+                return await listener(*args, **kwargs)
+            else:
+                log.debug(f"Guarded {listener.__qualname__} from invoking in {current_month.name}")
+        return guarded_listener
+    return decorator
+
+
+def in_month_command(*allowed_months: Month) -> t.Callable:
+    """
+    Check whether the command was invoked in one of `enabled_months`.
+
+    Uses the current UTC month at the time of running the predicate.
+    """
+    async def predicate(ctx: Context) -> bool:
+        current_month = resolve_current_month()
+        can_run = current_month in allowed_months
+
+        human_months = ", ".join(m.name for m in allowed_months)
+        log.debug(
+            f"Command '{ctx.command}' is locked to months {human_months}. "
+            f"Invoking it in month {current_month} is {'allowed' if can_run else 'disallowed'}."
+        )
+        if can_run:
+            return True
+        else:
+            raise InMonthCheckFailure(f"Command can only be used in {human_months}")
+
+    return commands.check(predicate)
+
+
+def in_month(*allowed_months: Month) -> t.Callable:
+    """
+    Universal decorator for season-locking commands and listeners alike.
+
+    This only serves to determine whether the decorated callable is a command,
+    a listener, or neither. It then delegates to either `in_month_command`,
+    or `in_month_listener`, or raises TypeError, respectively.
+
+    Please note that in order for this decorator to correctly determine whether
+    the decorated callable is a cmd or listener, it **has** to first be turned
+    into one. This means that this decorator should always be placed **above**
+    the d.py one that registers it as either.
+
+    This will decorate groups as well, as those subclass Command. In order to lock
+    all subcommands of a group, its `invoke_without_command` param must **not** be
+    manually set to True - this causes a circumvention of the group's callback
+    and the seasonal check applied to it.
```

Comment:
Yes, this could be added to the `if isinstance(callable_, Command):` clause:
```python
if hasattr(callable_, "invoke_without_command") and callable_.invoke_without_command:
```
However, I'm not sure how to handle such a situation. A warning-level log probably wouldn't be good - there currently exists [one place in the code base where the param is set to True and the decorator is used](https://github.com/python-discord/seasonalbot/blob/83f5b0c7b62df74edc0e58b67f535bfb11284c26/bot/exts/halloween/hacktoberstats.py#L35). I believe the param has to be there in this case, due to the way the group is designed - it takes a `github_username` argument, and should only be invoked if no sub-command was found.

To get around this, I simply decorated the group and both its sub-commands separately. It's not ideal, but I couldn't really think of anything more elegant, given the group's current design.

Reply:
Let's just leave it alone then.